software/main.py

```diff
@@ -0,0 +1,248 @@
+                if len(processedData.balls) > 0:
+                    state = "ball_move"
+                    continue
+                else:
+                    robot.move(0, 0, 6)
+            # End of ball_search
+            
+            # Moving towards ball
+            elif state == "ball_move":
+                LOGSTATE("ball_move")
+                if len(processedData.balls) > 0:
+                    # Movement logic
+                    speed_x = 0
+                    speed_y = 0
+                    speed_r = 0
+
+                    # Choosing the closest ball
+                    interesting_ball = processedData.balls[-1]
+                    #print("Ball:", interesting_ball)
+
+                    if interesting_ball.x < middle_x + 25 and interesting_ball.x > middle_x - 25 and interesting_ball.distance <= 475:
+                        state = "ball_orbit"
+                        basket_ok_counter = 0
+                        continue
+                    else:
+                        if interesting_ball.x > middle_x + 2 or interesting_ball.x < middle_x - 2:
+                            speed_x = sigmoid_controller(interesting_ball.x, middle_x, x_scale=2000, y_scale=max_speed)
+                            speed_r = -sigmoid_controller(interesting_ball.x, middle_x, x_scale=1200, y_scale=max_speed)
+                        if interesting_ball.distance > ball_good_range:
+                            speed_y = sigmoid_controller(interesting_ball.distance, ball_good_range, x_scale=1500, y_scale=max_speed)
+                        print(f"x: {speed_x}, y: {speed_y}, r: {speed_r}, dist: {interesting_ball.distance}, b.x: {interesting_ball.x}, b.y: {interesting_ball.y}")
+                        robot.move(speed_x, speed_y, speed_r)
+                else:
+                    state = "ball_search"
+                    continue
+            # End of ball_move
+
+            # Orbiting around ball until correct basket is found
+            elif state == "ball_orbit":
+                LOGSTATE("ball_orbit")
+                
+
+                if len(processedData.balls) > 0:
+                    interesting_ball = processedData.balls[-1]
+
+                    # For checking if the ball is still in position
+                    if interesting_ball.distance > 550:
+                        LOGE("Invalid radius, radius: " + str(interesting_ball.distance))
+                        state = "wait"
+                        continue
+
+                    # Determining the correct basket
+                    if using_magenta:
+                        basket = processedData.basket_m
+                    else:
+                        basket = processedData.basket_b
+
+                    if basket.exists:
+                        print("Basket x:", basket.x, "/", middle_x)
+                        #if (processedData.basket_m.x > (middle_x + 1) or processedData.basket_m.x < (middle_x - 1)):
+                        if (basket.distance < 1300):
```

Comment:
I would use a formula here. Something like `max((1.0 - (distance/max_distance)) * max_tolerance, min_tolerance)`

Reply:
This was temporary and is now removed.